Library/Tasks/ManagerTask.php

```diff
@@ -32,43 +33,63 @@ protected function initWorker(array $paramHash)
         // xxx factorise the topicname "build" logic
         $this->topicName = 'disturb-' . $this->workflowConfig['name'] . '-manager';
     }
+
     protected function processMessage(Dtos\Message $messageDto)
     {
         $this->getDI()->get('logger')->info('messageDto : ' . $messageDto);
-        $status = $this->workflowManagerService->getStatus($messageDto->getContract());
-        $this->getDI()->get('logger')->debug("Contract {$messageDto->getContract()} is '$status'");
         switch($messageDto->getType()) {
-        case Dtos\Message::TYPE_WF_CTRL:
-            switch($messageDto->getAction()) {
-            case 'start':
-                $this->workflowManagerService->init($messageDto->getContract());
-                $this->runNextStep($messageDto->getContract());
+            case Dtos\Message::TYPE_WF_CTRL:
+                switch($messageDto->getAction()) {
+                    case 'start':
+                        $this->workflowManagerService->init($messageDto->getId());
+                        $this->runNextStep($messageDto->getId());
+                    break;
+                }
                 break;
-            }
-            break;
-        case Dtos\Message::TYPE_STEP_ACK:
-            $this->getDI()->get('logger')->debug("Step {$messageDto->getStep()} says {$messageDto->getResult()}");
-            $stepResultHash = json_decode($messageDto->getResult(), true);
+            case Dtos\Message::TYPE_STEP_ACK:
+                $this->getDI()->get('logger')->debug("Step {$messageDto->getStepCode()} says {$messageDto->getResult()}");
+                $stepResultHash = json_decode($messageDto->getResult(), true);
+                $this->workflowManagerService->processStepJobResult(
+                    $messageDto->getId(),
+                    $messageDto->getStepCode(),
+                    $messageDto->getJobId(),
+                    $stepResultHash
+                );
 
-            $step = $this->workflowManagerService->finalizeStep($messageDto->getContract(), $messageDto->getStep(), $stepResultHash);
-            $this->runNextStep($messageDto->getContract());
-            break;
-        default :
-            $this->getDI()->get('logger')->error("ERR : Unknown message type : {$messageDto->getType()}");
+                $status = $this->workflowManagerService->getStatus($messageDto->getId());
+                $this->getDI()->get('logger')->debug("Id {$messageDto->getId()} is '$status'");
+                if ($status == Services\WorkflowManager::STATUS_FAILED) {
+                    throw new WorkflowException("Id failed {$messageDto->getId()}");
+                }
+
+                if ($this->workflowManagerService->isNextStepRunnable($messageDto->getId()))
```

Comment:
c.f. didCurrentStepSucceeded comment

execflow should looks like this
```
 - getCurrentStepStatus
   - Success : hasNextStep
      - Y : runNextStep
      - N : finalizeWorkflowExec
   - Failed : processStepFailure
   - Running : noop
```
then we must handle the "end of the workflow execution"